### PR TITLE
PM-1377 Calling Revert when patch fails - VPOS-XPAY

### PR DESCRIPTION
**Pull request by DavideAmatoNexi**
#### List of Changes

Calling Revert When patch fails - **VPOS &  XPAY**

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have created or updated an API interface.
  - [ ] I have updated the OpenAPI spec
- [ ] I have created or updated a database table
  - [ ] I have updated both the Oracle and Postgres migrations
- [ ] My PR is cohesive and self-contained (rule of thumb: less than 250 changed lines) 

### Link to story

https://pagopa.atlassian.net/browse/PM-1377

### CODE REVIEW

File: src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java [hunk 1]
```diff
@@ -400,9 +400,22 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
...
     }
 
+    private void executeRevertAfterPatch(PaymentRequestEntity entity) {
+        String requestId = entity.getGuid();
+        try {
+            if (executeXPayOrderStatus(entity).equals(OK)) {
+                executeXPayRevert(entity);
+            }
+        } catch (Exception e) {
+            log.error(PATCH_CLOSE_PAYMENT_ERROR + requestId);
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java:414]:
```suggestion
            log.error("{}{}", PATCH_CLOSE_PAYMENT_ERROR, requestId);
```

File: src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java [hunk 2]
```diff
@@ -400,9 +400,22 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
...
+        }
+        log.info("End Refunding payment with requestId: " + requestId);
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java:416]:
```suggestion
        log.info("End Refunding payment with requestId: {}", requestId);
```

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java [hunk 3]
```diff
@@ -117,8 +119,8 @@ private void executeStep1(Map<String, String> params, PaymentRequestEntity entit
             }
 
             //If the resultCode is 26, the PATCH is not called
-            if(!"26".equals(response.getResultCode())) {
-                executePatchTransaction(entity);
+            if (!"26".equals(response.getResultCode())) {
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java:122]: Refactor `"26"` into a constant field please :)

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java [hunk 4]
```diff
@@ -226,6 +228,54 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
...
             log.error("{}{}", PATCH_CLOSE_PAYMENT_ERROR, requestId, e);
             log.info("Refunding payment with requestId: {}", requestId);
+            if (executeOrderStatus(entity, pgsRequest).equals(OK)) {
+                executeRevert(entity, pgsRequest);
+            }
+        }
+    }
+
+    private RefundOutcome executeOrderStatus(PaymentRequestEntity entity, StepZeroRequest stepZeroRequest) throws IOException {
+        log.info("Calling VPOS - OrderStatus - for requestId: " + entity.getGuid());
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java:238]:
```suggestion
        log.info("Calling VPOS - OrderStatus - for requestId: {}", entity.getGuid());
```

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java [hunk 5]
```diff
@@ -226,6 +228,54 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
...
+        } else {
+            entity.setErrorMessage("Error during orderStatus");
+            paymentRequestRepository.save(entity);
+            return KO;
+        }
+    }
+
+    private void executeRevert(PaymentRequestEntity entity, StepZeroRequest pgsRequest) {
+        try {
+            log.info("Calling VPOS - Revert - for requestId: " + entity.getGuid());
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java:258]:
```suggestion
            log.info("Calling VPOS - Revert - for requestId: {}", entity.getGuid());
```

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java [hunk 6]
```diff
@@ -226,6 +228,54 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
...
+            Map<String, String> params = vPosRequestUtils.buildRevertRequestParams(pgsRequest, entity.getCorrelationId());
+            HttpClientResponse clientResponse = callVPos(params);
+            AuthResponse response = vPosResponseUtils.buildAuthResponse(clientResponse.getEntity());
+            vPosResponseUtils.validateResponseMac(response.getTimestamp(), response.getResultCode(), response.getResultMac(), pgsRequest);
+            checkRevertResultCode(response, entity);
+        } catch (Exception e) {
+            log.error(GENERIC_REFUND_ERROR_MSG + entity.getIdTransaction() + " cause: " + e.getCause() + " - " + e.getMessage(), e);
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java:265]:
```suggestion
            log.error("{}{} cause: {} - {}", GENERIC_REFUND_ERROR_MSG, entity.getIdTransaction(), e.getCause(),  e.getMessage(), e);
```

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java [hunk 7]
```diff
@@ -226,6 +228,54 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
...
+        String resultCode = response.getResultCode();
+        if (resultCode.equals(RESULT_CODE_AUTHORIZED)) {
+            entity.setStatus(CANCELLED.name());
+            entity.setIsRefunded(true);
+            paymentRequestRepository.save(entity);
+        } else {
+            entity.setErrorMessage("Error during Revert");
+            entity.setIsRefunded(false);
         }
+        log.info("END - VPos Request Payment Revert for requestId " + entity.getGuid() + " - resultCode: " + resultCode);
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java:279]:
```suggestion
        log.info("END - VPos Request Payment Revert for requestId {}  - resultCode: {}" + entity.getGuid(), resultCode);
```

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java [hunk 8]
```diff
@@ -202,6 +205,54 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
...
             log.error("{}{}", PATCH_CLOSE_PAYMENT_ERROR, requestId, e);
             log.info("Refunding payment with requestId: {}", requestId);
+            if (executeOrderStatus(entity, pgsRequest).equals(OK)) {
+                executeRevert(entity, pgsRequest);
+            }
+        }
+    }
+
+    private RefundOutcome executeOrderStatus(PaymentRequestEntity entity, StepZeroRequest stepZeroRequest) throws IOException {
+        log.info("Calling VPOS - OrderStatus - for requestId: " + entity.getGuid());
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java:215]:
```suggestion
        log.info("Calling VPOS - OrderStatus - for requestId: {}", entity.getGuid());
```

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java [hunk 9]
```diff
@@ -202,6 +205,54 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
...
+    private void executeRevert(PaymentRequestEntity entity, StepZeroRequest pgsRequest) {
+        try {
+            log.info("Calling VPOS - Revert - for requestId: " + entity.getGuid());
+            Map<String, String> params = vPosRequestUtils.buildRevertRequestParams(pgsRequest, entity.getCorrelationId());
+            HttpClientResponse clientResponse = callVPos(params);
+            AuthResponse response = vPosResponseUtils.buildAuthResponse(clientResponse.getEntity());
+            vPosResponseUtils.validateResponseMac(response.getTimestamp(), response.getResultCode(), response.getResultMac(), pgsRequest);
+            checkRevertResultCode(response, entity);
+        } catch (Exception e) {
+            log.error(GENERIC_REFUND_ERROR_MSG + entity.getIdTransaction() + " cause: " + e.getCause() + " - " + e.getMessage(), e);
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java:242]:
```suggestion
            log.error("{}{} cause: {} - {}", GENERIC_REFUND_ERROR_MSG, entity.getIdTransaction(), e.getCause(),  e.getMessage(), e);
```

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java [hunk 10]
```diff
@@ -202,6 +205,54 @@ private void executePatchTransaction(PaymentRequestEntity entity) {
...
+        if (resultCode.equals(RESULT_CODE_AUTHORIZED)) {
+            entity.setStatus(CANCELLED.name());
+            entity.setIsRefunded(true);
+            paymentRequestRepository.save(entity);
+        } else {
+            entity.setErrorMessage("Error during Revert");
+            entity.setIsRefunded(false);
         }
+        log.info("END - VPos Request Payment Revert for requestId " + entity.getGuid() + " - resultCode: " + resultCode);
     }
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java:257]: This function has a duplicate in `CcResumeStep1Service.java`

File: src/main/java/it/pagopa/pm/gateway/service/VposService.java [hunk 11]
```diff
@@ -301,8 +308,29 @@ private void checkRevertResultCode(AuthResponse response, PaymentRequestEntity e
...
+        log.info("END - VPos Request Payment Revert for requestId " + entity.getGuid() + " - resultCode: " + resultCode);
+    }
+
+    private RefundOutcome executeOrderStatus(PaymentRequestEntity entity, StepZeroRequest stepZeroRequest) throws IOException {
+        log.info("Calling VPOS - OrderStatus - for requestId: " + entity.getGuid());
+        Map<String, String> params = vPosRequestUtils.buildOrderStatusParams(stepZeroRequest);
+        HttpClientResponse clientResponse = callVPos(params);
+        VposOrderStatusResponse response = vPosResponseUtils.buildOrderStatusResponse(clientResponse.getEntity());
+        return computeOrderStatusResultCode(response, entity);
     }
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/VposService.java:324]: Duplicate in `CcResumeStep1Service`

**Review by giovanniberti — CHANGES_REQUESTED** [review]

File: src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java (hunk 1 above)
Comment (DavideAmatoNexi) [reply src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java]: https://github.com/pagopa/pagopa-payment-transactions-gateway/pull/239/commits/6da10a916222aeb83d80e8dd48d1f471c7508b4e

File: src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java (hunk 2 above)
Comment (DavideAmatoNexi) [reply src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java]: https://github.com/pagopa/pagopa-payment-transactions-gateway/pull/239/commits/29a20d25d62bd336cbe45dc5f1cd5cc016116b89

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java (hunk 5 above)
Comment (DavideAmatoNexi) [reply src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java:258]: https://github.com/pagopa/pagopa-payment-transactions-gateway/pull/239/commits/5e541eda0c2dcbcef461510fc0757354cc069321

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java (hunk 6 above)
Comment (DavideAmatoNexi) [reply src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java]: https://github.com/pagopa/pagopa-payment-transactions-gateway/pull/239/commits/b52f46088d2ec08cbc45983434881b22015268b1

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java (hunk 3 above)
Comment (DavideAmatoNexi) [reply src/main/java/it/pagopa/pm/gateway/service/CcResumeStep1Service.java]: https://github.com/pagopa/pagopa-payment-transactions-gateway/pull/239/commits/7035740f5c2b7c23029194ae1fa443550e0e1050 Done

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java (hunk 10 above)
Comment (DavideAmatoNexi) [reply src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java]: https://github.com/pagopa/pagopa-payment-transactions-gateway/commit/7035740f5c2b7c23029194ae1fa443550e0e1050 I did a refactoring

**Review by infantesimone — APPROVED** [review]

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java [hunk 12]
```diff
@@ -66,9 +61,7 @@ public class CcResumeStep2Service {
     private ObjectMapper objectMapper;
 
     @Autowired
-    public CcResumeStep2Service(ClientsConfig clientsConfig) {
-        this.clientsConfig = clientsConfig;
-    }
+    private VposPatchUtils vposPatchUtils;
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java:64]: Prefer constructor injection to field injection

File: src/main/java/it/pagopa/pm/gateway/service/VposService.java [hunk 13]
```diff
@@ -301,8 +270,11 @@ private void checkRevertResultCode(AuthResponse response, PaymentRequestEntity e
             entity.setStatus(CANCELLED.name());
             entity.setIsRefunded(true);
             paymentRequestRepository.save(entity);
+        } else {
+            entity.setErrorMessage("Error during Revert");
+            entity.setIsRefunded(false);
         }
-        log.info("END - VPos Request Payment Revert for requestId " + entity.getGuid());
+        log.info("END - VPos Request Payment Revert for requestId " + entity.getGuid() + " - resultCode: " + resultCode);
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/VposService.java:277]:
```suggestion
        log.info("END - VPos Request Payment Revert for requestId {} - resultCode: {}", entity.getGuid(), resultCode);
```

Comment (DavideAmatoNexi) [reply src/main/java/it/pagopa/pm/gateway/service/VposService.java]: Done https://github.com/pagopa/pagopa-payment-transactions-gateway/pull/239/commits/0277898eacaea3e6111025aac69ca21c03452e68

File: src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java (hunk 12 above)
Comment (DavideAmatoNexi) [reply src/main/java/it/pagopa/pm/gateway/service/CcResumeStep2Service.java]: Done https://github.com/pagopa/pagopa-payment-transactions-gateway/pull/239/commits/d7187758c5845f94735a249624d783b9e9531754

File: src/main/java/it/pagopa/pm/gateway/utils/VposPatchUtils.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+            entity.setErrorMessage("Error during Revert");
+            entity.setIsRefunded(false);
+        }
+        log.info("END - VPos Request Payment Revert for requestId {}  - resultCode: {}" + entity.getGuid(), resultCode);
+    }
+
+    private HttpClientResponse callVPos(Map<String, String> params) throws IOException {
+        HttpClientResponse clientResponse = httpClient.post(vposUrl, ContentType.APPLICATION_FORM_URLENCODED.getMimeType(), params);
+        if (clientResponse.getStatus() != HttpStatus.OK.value()) {
+            log.error("HTTP Response Status: " + clientResponse.getStatus());
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/utils/VposPatchUtils.java:142]:
```suggestion
            log.error("HTTP Response Status: {}", clientResponse.getStatus());
```

Comment (DavideAmatoNexi) [reply src/main/java/it/pagopa/pm/gateway/utils/VposPatchUtils.java]: https://github.com/pagopa/pagopa-payment-transactions-gateway/pull/239/commits/6765cf50dafffaca92d3d3f049fe514d030e8011

**Review by giovanniberti — APPROVED** [review]

**Pull request merged**